samcli/lib/pipeline/bootstrap/resource.py

```diff
@@ -0,0 +1,133 @@
+""" Represents AWS resource"""
+from typing import Optional
+
+
+class ARNParts:
+    """
+    Decompose a given ARN into its parts https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html
+
+    Attributes
+    ----------
+    partition: str
+        the partition part(AWS, aws-cn or aws-us-gov)  of the ARN
+    service: str
+        the service part(S3, IAM, ECR, ...etc) of the ARN
+    region: str
+        the AWS region part(us-east-1, eu-west-1, ...etc) of the ARN
+    account-id: str
+        the account-id part of the ARN
+    resource-id: str
+        the resource-id part of the ARN
+    resource-type: str
+        the resource-type part of the ARN
+    """
+    partition: str
+    service: str
+    region: str
+    account_id: str
+    resource_id: str
+
+    def __init__(self, arn: str) -> None:
+        if not isinstance(arn, str):
+            raise ValueError(f"Invalid ARN ({arn}) is not a String")
```

Comment:
ARN is a user input so I made it that way just in case if the user entered non-string input (numbers for example) and we didn't cast to string. So basically a user can enter a number and python don't enforce types.

Reply:
That makes sense. However, it is confusing here and I don't think it is `Resource`'s job to deal with it. The user input for `ARN` should be converted to string when we receive the inputs, maybe in `GuidedContext`. 

I did a test and looks like `click.prompt()` convert the input to string itself, see:
<img width="672" alt="image" src="https://user-images.githubusercontent.com/3804518/115446773-8adbe800-a1cc-11eb-84bf-cc1f93a6353c.png"> Do we really need this check at all?

Reply:
this is because we specifically set the type to `click.STRING`. I'm just saying what if we forgot to make such a casting. This check just adds one more layer of safety in this case, otherwise, it will raise an AttributeError, and every call to this method should handle both ValueError and AttributeError instead of ValueError only.

Reply:
> I'm just saying what if we forgot to make such a casting. 

For now we don't need to worry about it. Even if we forgot to make such a casting happens in the future, it is our code bug that we should fix. Besides from a code comprehension aspect, it creates this extra mental burden: "if the type does not match the hint, how about others, should we do this everywhere?" I don't think we want to do this